specgen/specgen/main.go

```diff
@@ -0,0 +1,117 @@
+// Command specgen provides a utility for generating utls.ClientHelloSpecs. This command starts an
+// HTTPS server which responds to every request with a ClientHelloSpec corresponding to the TLS
+// ClientHello the server received. The spec is logged for every request as well. Make a request of
+// the server via web browser to see a ClientHelloSpec for the browser.
+//
+// It is recommended that mkcert be used to generate the keypair for this server. This ensures
+// browsers do not flag and block the page. See https://github.com/FiloSottile/mkcert.
+package main
```

Comment:
Shouldn't this be in `cmd/specgen`? Is there any utility to having `specgen` as a library? otherwise I'd put it all in `main`.

Reply:
This is an example of a leaky interface. Separation of concerns is great, but if all you end up doing is pushing the implementation details around between the concerns you haven't really achieved anything, in my opinion.

I think we've disagreed about this stuff before (when to split packages), so just to outline my process for this:

 - If I see something that can be cleanly split into two packages, great. I try to do that and see how the interface works out.
 - If I put my hand over my left eye, does module B make sense on its own? If I put my hand over my right eye, does module A make sense on its own?
 - If not, I go back and put them together again.
 - Later on, if it's causing lots of problems because the module is way too big, I start the same process  again, with the knowledge I've gained in the intervening time and a different threshold for "makes sense on its own". 

To me,`specgen` fails the hand-over-one-eye test. What is the package doing exactly? Could you explain it to a developer who only sees it and not the surrounding context? It seems extremely linked to this command line tool, and not generally useful. And it's not like this is a huge unwieldy package which needs to be broken apart because it's unreadable.

Reply:
Your definition of a leaky interface is great.

> To me,specgen fails the hand-over-one-eye test. What is the package doing exactly?

This is explained in the package doc comment: "Package specgen provides a utility for generating utls.ClientHelloSpec."  To elaborate, the package allows one to generate Go code (the `utls.ClientHelloSpec`) from a raw TLS ClientHello.

> It seems extremely linked to this command line tool, and not generally useful.

Agree to disagree, I guess?  This could be useful in writing tests, in spot checking already-defined `ClientHelloSpec`s, or in writing alternative forms of the `specgen` command-line tool (i.e. not a local server).

You and I recently discussed the possibility of some kind of hosted server for capturing hellos from clients.  The command-line tool would not suit this purpose, but `specgen` and `hellocap` could be composed in a new way to achieve this - I believe without changes to the packages as they currently exist.  I think this directly contradicts your hypothesis that this package is not generally useful.

Reply:
It's up to you, but on that last point, the server definitely wouldn't be spitting out the go code, would it? Dumping out a go struct definition seems only useful on the command line to me, where you can copy paste it.

Reply:
It will!  It's pretty easy to test this for yourself.  Here's what it looks like:

![image](https://user-images.githubusercontent.com/4043597/200736129-9e6a116d-3c8f-4de5-a9a7-b1e9a5cdb54f.png)

I found this more user-friendly than copy/pasting from the command-line.  The difference is pretty marginal to be honest.  But you have to hit the server from the browser anyway (since you're trying to capture the browser's hello).  So you're already on the page, then you just ctrl+a, ctrl+c.

Reply:
Sorry, I meant the hypothetical production server that you mentioned, not this server here (which I was arguing is tightly coupled with `specgen`).

Reply:
This is actually a bad example for my crusade against prematurely splitting packages though :) to your actual point, "generating a uTLS spec from a binary blob containing a clienthello" is a good enough description to stand on its own.